app/src/main/java/ch/epfl/sdp/drone3d/service/impl/drone/DroneExecutorImpl.kt

```diff
@@ -47,136 +50,174 @@ class DroneExecutorImpl(
-                .doOnComplete{ ToastHandler.showToastAsync(context, "Drone took off") }
-
-        return start(context, instance, missionPlan, takeoff)
-    }
-
-    private fun start(context: Context, instance: System, missionPlan: Mission.MissionPlan,
-                      before: Completable = Completable.complete()): Completable {
-
-        val startMission = before.doOnComplete { data.getMutableDroneStatus().postValue(STARTING_MISSION) }
-                .andThen(instance.mission.setReturnToLaunchAfterMission(true))
-                .andThen(instance.mission.uploadMission(missionPlan)
-                        .andThen(instance.mission.startMission())
-                        .doOnComplete {
-                            data.getMutableDroneStatus().postValue(EXECUTING_MISSION)
-                            data.getMutableMission().postValue(missionPlan.missionItems.dropLast(1))
-                            data.getMutableMissionPaused().postValue(false)
-                            ToastHandler.showToastAsync(context, R.string.drone_mission_success)
-                        })
-
-        return finish(context, instance, startMission)
-    }
-
-    private fun finish(context: Context, instance: System, before: Completable): Completable {
-        return before.andThen(instance.mission.missionProgress).distinctUntilChanged()
-                .filter{ it.current >= it.total - 1 }.firstOrError().toCompletable()
-                .doOnComplete{
-                    ToastHandler.showToastAsync(context, "Mission done, returning to launch point")
-                    data.getMutableDroneStatus().postValue(GOING_BACK)
-                }
-                .andThen(instance.mission.missionProgress).distinctUntilChanged()
-                .filter{ it.current == it.total }.firstOrError().toCompletable()
-                .doOnComplete{ data.getMutableDroneStatus().postValue(LANDING) }
-                .andThen(instance.action.land())
-                .andThen(instance.action.disarm())
-                .doOnComplete{
-                    // Completion
-                    data.getMutableDroneStatus().postValue(IDLE)
-                    data.getMutableMission().postValue(null)
-                    data.getMutableMissionPaused().postValue(true)
-                }
+    private fun start(ctx: Context, instance: System, missionPlan: Mission.MissionPlan): Completable = 
+            Completable.fromCallable { data.getMutableDroneStatus().postValue(SENDING_ORDER) }
+                    .andThen(instance.mission.setReturnToLaunchAfterMission(false))
+                    .andThen(instance.mission.uploadMission(missionPlan)
+                            .andThen(instance.mission.startMission())
+                            .doOnComplete {
+                                data.getMutableDroneStatus().postValue(EXECUTING_MISSION)
+                                data.getMutableMission().postValue(missionPlan.missionItems.dropLast(1))
```

Comment:
I don't understand why the last element is dropped here.

Reply:
So, that's a special thing.
I want the drone to come back to its launch point by default.

But I also want to be able to cancel a mission saying it should come back to its launch point or to the user.

Coming back to launch after the end of the mission is possible.
But as there is also the case where the drone comes back to the user, I removed that possibility and made the launch point (or the user location) the last point of the mission so that the drone should land when the mission is finish either if it is on the home location or the user's location.

So, as the last point is in the drone mission but not actually part of the mission, I drop it from the live data.